Bug fix to ensure that calibrations are carried over by circuit compose

eggerdj opened the pull request:
### Summary
Fixes #5394 

### Details and comments
* The way the `_calibrations` dict is updated has been changed.
* A corresponding unit test has been added.

lcapelluto commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -681,7 +681,12 @@ def compose(self, other, qubits=None, clbits=None, front=False, inplace=False):
             dest._update_parameter_table(instr)
-        dest._calibrations.update(other.calibrations)
+
+        for gate, cals in other.calibrations.items():
+            if gate in dest._calibrations:
+                dest._calibrations[gate].update(cals)
+            else:
+                dest._calibrations[gate] = cals
```
There's actually another bug I meant to fix:
```
--- a/qiskit/converters/dag_to_circuit.py
+++ b/qiskit/converters/dag_to_circuit.py
@@ -11,6 +11,8 @@
 # that they have been altered from the originals.

 """Helper function for converting a dag to a circuit."""
+from collections import defaultdict
+
 from qiskit.circuit.quantumcircuit import QuantumCircuit


@@ -48,7 +50,8 @@ def dag_to_circuit(dag):
     name = dag.name or None
     circuit = QuantumCircuit(*dag.qregs.values(), *dag.cregs.values(), name=name,
                              global_phase=dag.global_phase)
-    circuit.calibrations = dag.calibrations
+    circuit._calibrations = defaultdict(dict)
+    circuit._calibrations.update(dag.calibrations)

     for node in dag.topological_op_nodes():
         # Get arguments for classical control (if any)
```

After fixing that, this should be:
```suggestion
        for gate, cals in other.calibrations.items():
            dest._calibrations[gate].update(cals)
```

eggerdj replied:
Done.

ajavadia reviewed (commented):
does this need to be in other converters as well? circuit <-> dag <-> dependencydag
During transpiler all 3 may be converted to each other and I don't think the cals should get lost

eggerdj commented:
@ajavadia Some of the conversion were not doing this properly. I added a fix as well as a couple of tests to catch this.

lcapelluto commented:
Can you add this test:

```
qc = QuantumCircuit(2, 2)
qc.h(0)
qc.cx(0, 1)
qc.measure(0, 0)
qc = transpile(qc, backend)
qc.add_calibration('cx', [0, 1], pulse.Schedule())
```
With current master, this fails because the transpiled circuit `_calibrations` is not a default dict, and it raises a `KeyError` on `'cx'`.

eggerdj commented:
Thanks @lcapelluto . There was some converting of the `defaultdict` to `dict` in the properties of `DagCircuit` and `DagDependency` which was causing `qc.add_calibration('cx', [0, 1], pulse.Schedule())` to fail. I removed the dict conversion and added the test you suggested.

lcapelluto commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -205,12 +205,12 @@ def data(self):
             {'gate_name': {(qubits, params): schedule}}
         """
-        return dict(self._calibrations)
+        return self._calibrations
```
With this implementation, something like this:
```
cals = qc.calibrations
dne_gates = cals['DNE']
```
would return an empty dict rather than raising a `KeyError`. This isn't that bad, and trying to grab any cal from `dne_gates` would then fail, but it's strange to build a `calibrations` which contains all the keys for which you've tried to key into.

We could update the setter below to turn the input into a `defaultdict(dict)`, and then the other files would stay the same. 

I'm not as concerned about returning `self._calibrations` from the dag, since I believe it's not as often directly accessed by users
